Cleanup code that uses MCStringGetCString(), part 1

peter-b opened the pull request:
`MCStringGetCString()` is problematic because it silently forces any string that it's called for to native representation, even if that results in data loss.

This patch refactors some hairy code that uses `MCStringGetCString()` unnecessarily.

peter-b committed:
Avoid MCStringGetCString() in Windows crash dump generation

This enables the creation of crash dump files that have special
characters in the path.

livecodeali commented on engine/src/dskw32.cpp:
```diff
@@ -3515,17 +3515,12 @@ struct MCWindowsDesktop: public MCSystemInterface, public MCWindowsSystemService
 			MCAutoStringRef t_string;
 			/* UNCHECKED */ MCStringCreateWithCString(*t_utf8, &t_string);
-			t_environment -> Run(*t_string, &t_result);
+			t_environment -> Run(p_string, &t_result);
```
I think you mean `p_script` here

livecodeali replied:
And I guess you can delete the other lines...

livecodeali commented:
I'm not sure this will work with `MCWindowsActiveScriptEnvironment::Run` - doesn't that need to change so that it uses an `MCAutoStringRefAsSysString` for the incoming script, and explicitly create a string from utf-8 chars from the result on the way out, rather than using `MCStringCreateWithCString`?

peter-b commented:
@livecodeali Thanks for looking over this -- yeah, I need to go back over this and make some revisions. :-/

runrevmark commented on engine/src/w32script.cpp:
```diff
@@ -596,10 +596,9 @@ void MCWindowsActiveScriptEnvironment::Finalize(void)
 void MCWindowsActiveScriptEnvironment::Run(MCStringRef p_script, MCStringRef& r_out)
 {
 	LPOLESTR t_ole_script;
```
I'm pretty sure LPOLESTR is the same as LPWSTR - so I think we can use MCAutoStringRefAsWString here. I think the reason that LPOLESTR is used in this case (in the Win32 API) is that the type of strings used in OLE were always Unicode - whereas other Win32 APIs would be Ansi or Unicode.

runrevmark commented:
Btw, the only reason the scriptenv stuff used UTF-8 was to ensure unicode could be communicated to the scripts which are run (as it was implemented to support network proxy pac files) - I don't think going through UTF-8 is needed now.

runrevmark commented:
@peter-b: Removal of the CString getters is definitely what we want to do, however it leaves a small hole... I use GetCString all the time to show values in the debugger in Xcode / VS (not that it is ideal for that!) so some sort of debugger friendly function for dumping a valueref would be useful - perhaps a pair - one which dumps using the standard debug log facility, and one which returns a c-string in appropriate encoding, but stores the c-string in a global so it only lasts until the next call to the function. (I was pondering whether that would need to be a thread-local, but this would be for debugging only, so human invoked).

peter-b committed:
MCS_posturl_callback(): Use MCStringEncode to create data from string

peter-b committed:
Keep MCWindowsActiveScriptEnvironment's UTF-8 as impl detail

The implementation of running alternate language scripts on Windows
was going through some gymnastics to create strings that thought they
were native strings while actually containing UTF-8 data, i.e. it was
creating strings containing deliberately-corrupted data.

This was required by `MCWindowsActiveScriptEnvironment::Run()`, which
was accepting "native" strings and then treating them as though they
contained UTF-8.

This patch modifies `MCWindowsActiveScriptEnvironment::Run()` to keep
its use of UTF-8 encoding as an implementation detail, and use
"normal" strings in its interface.

peter-b commented:
@ali @runrevmark I've addressed the feedback on this -- thank you for reviewing it! The changes since the original version of the patch were:

- add `MCStringCreateWithBSTR()` in order to enable the complete removal of UTF-8 intermediate representation from `MCWindowsActiveScriptEnvironment::Run()` -- thanks to @runrevmark for explaining how `BSTR` works!
- cleaned up some additional unnecessary code (thanks @livecodeali)
- added a necessary 🚨 `const_cast`🚨  in the Android URL-related stuff

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -6820,6 +6820,25 @@ __MCStringCreateWithStrings(MCStringRef& r_string, bool p_has_separator, unichar
+MCStringCreateWithBSTR(const BSTR p_bstr,
+                       MCStringRef& r_string)
+{
+    return MCStringCreateWithBytes(reinterpret_cast<const byte_t*>(p_bstr),
```
Would it not be more appropriate to use CreateWithChars here? BSTR's are UTF-16 strings so are actually arrays of uint16_t, not bytes.

peter-b committed:
libfoundation: Add MCStringCreateWithBSTR()

`BSTR` is a string type used by COM, Automation and Interop functions
in the Windows API.

peter-b commented:
@runrevmark I've switched `MCStringCreateWithBSTR()` to use `MCStringCreateWithChars()` internally.

runrevmark commented:
@livecode-vulcan review ok d8bf081

livecode-vulcan commented:
💙 review by @runrevmark ok d8bf0812b8c631ed6691383426d34e279b0926b8

livecode-vulcan commented:
😎 test success d8bf0812b8c631ed6691383426d34e279b0926b8

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/119)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/119)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/105)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/51)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/51)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/117)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/119)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/121)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/123)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/121)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/123)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/139)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/118)
